### PR TITLE
balancerd: use dycnfgs and launch darkly

### PR DESCRIPTION
**Pull request by maddyblue**
Add a new crate to sync launchdarkly to a configset. Hook it up to the balancer.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/dyncfg-ld/src/lib.rs
Comment (maddyblue) [inline src/dyncfg-ld/src/lib.rs:25]: Having this be the only `pub` thing in the crate I think logically follows from how it works (it used to return a struct but...there was no point). But I don't love the API of "call this function, it returns ~nothing and probably syncs are being done". Thoughts? Is it fine? Other ideas?

**Comment by jkosh44** [comment]
Balancer stuff LGTM, but I had the following questions/comments:
- Dan should probably look at the dyncfg stuff because I'm not that familiar.
- Do we need to worry about balancer flag names colliding with adapter flag names, or are they namespaced somehow?
- Should we be using the new crate/code in the adapter as well as in the balancer?

File: src/dyncfg-ld/src/lib.rs
Comment (danhhz) [reply src/dyncfg-ld/src/lib.rs:25]: This is the API I'd have expected before seeing the PR, so seems fine to me.

If you really wanted, you could hand back the sync loop future (or `SyncedConfigSet` with a `pub async fn sync_loop`) and make the caller put it into a task.

File: src/balancerd/src/service.rs
Comment (danhhz) [inline src/balancerd/src/service.rs:92]: There's a big comment in envd explaining that we require an LD sync the first time an env starts (because the defaults in code could be arbitrarily out of date) but that we allow timing out after the first boot (so that LD isn't a tier 1 dep and whatever was stored in the catalog from the previous boot is probably fine). https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/coord.rs#L3816-L3854

Of course, there's no catalog in balancerd, so we can't do the same. The policy you have in the PR seems fine to me, but could be getting a quick @benesch feels on the tradeoffs.

File: src/dyncfg-ld/src/lib.rs
Comment (danhhz) [inline src/dyncfg-ld/src/lib.rs:92]: If it's easy to share code for this, I'd do it. E.g. make `ld_ctx` pub in this crate and call it from envd. Seems like the type of thing that could easily rot in one codepath.

File: src/dyncfg-ld/src/lib.rs
Comment (danhhz) [inline src/dyncfg-ld/src/lib.rs:145]: Could store a map of name -> ld::FlagValue, update it at the end of a sync, and use that here to diff what's changed

**Review by danhhz — APPROVED** [review]
Oh hell yes, I've been wanting this for persistcli!

LGTM, left some notes but they're all optional.

File: src/dyncfg-ld/Cargo.toml
Comment (benesch) [inline src/dyncfg-ld/Cargo.toml:2]: I feel like it's worth fully spelling this out as `mz-dyncfg-launchdarkly`, for consistency with the other (official) `launchdarkly` crates.

File: src/balancerd/src/service.rs
Comment (benesch) [reply src/balancerd/src/service.rs:92]: Seems fine to me too!

File: src/balancerd/src/lib.rs
Comment (benesch) [inline src/balancerd/src/lib.rs:175]: Can you configure the context with a new `cloud_provider_region` context that sets `cloud_provider` and `cloud_provider_region` keys, sourced from CLI flags? Would be nice to be able to target only balancers in a certain cloud provider region.

**Review by benesch — COMMENTED** [review]
This is awesome, thanks! Dropped in some very high level feedback but leaving the in-depth review to @danhhz.

File: src/dyncfg-ld/src/lib.rs
Comment (benesch) [reply src/dyncfg-ld/src/lib.rs:25]: Agree, also the API I'd have expected.

**Comment by benesch** [comment]
> Should we be using the new crate/code in the adapter as well as in the balancer?

I think long term ideal would be to share, but the adapter code is weird and special enough with its durability requirements that I'm totally comfortable punting on that for now.

**Comment by maddyblue** [comment]
> Do we need to worry about balancer flag names colliding with adapter flag names, or are they namespaced somehow?

They are not namespaced and we do need to worry about collision. Prefixing with `balancer_` should be ok.

File: src/balancerd/src/service.rs
Comment (maddyblue) [reply src/balancerd/src/service.rs:92]: Added a buncha comments scaring us into choosing good default values so if LD is down we still start and things are reasonable.

File: src/dyncfg-ld/Cargo.toml
Comment (maddyblue) [reply src/dyncfg-ld/Cargo.toml:2]: Done.

File: src/dyncfg-ld/src/lib.rs
Comment (maddyblue) [reply src/dyncfg-ld/src/lib.rs:92]: There's already an existing ld_ctx fn in the adapter crate that does specific stuff for an Environment. It's not obviously shareable to me. I think it's fine for the current PR.

File: src/dyncfg-ld/src/lib.rs
Comment (maddyblue) [reply src/dyncfg-ld/src/lib.rs:145]: I looked at the adapter code again and it seems that it updates everything each time. I'm going to leave as is.

File: src/balancerd/src/lib.rs
Comment (maddyblue) [inline src/balancerd/src/lib.rs:182]: I added cloud provider and region entries here, but I think I'm doing it wrong. Here it's being added as a top-level kind, one for provider and one for region. But the adapter merges them together in a higher entry: https://github.com/MaterializeInc/materialize/blob/e8c17e0285b9fa313ea337593ca22eb75345821c/src/adapter/src/config/frontend.rs#L159

I don't grok LD enough to understand which one I should be using. Suggestions? Fine as is?

**Comment by maddyblue** [comment]
@alex-hunt-materialize There's a buncha CLI changes here. Does the cloud team need this PR blocked until ya'll have something in place? Or do you pin the balancerd version to a specific SHA, so we can safely merge this change for now?

**Comment by alex-hunt-materialize** [comment]
The balancerd image is chosen based on the environmentd image. We will need to update any changed cli args before updating the environmentd image to any built after this change.

File: src/balancerd/src/lib.rs
Comment (benesch) [reply src/balancerd/src/lib.rs:182]: I'd do it like this:

```
    if let Some(provider) = cfg.cloud_provider.clone() {
            ld::ContextBuilder::new(provider)
                .kind("cloud-provider")
                .set_string("cloud_provider_region", cfg.cloud_provider_region.clone().unwrap_or("unknown"))
    }
```

Idea is that we want only one context kind (`cloud-provider`), and we attach the region if it's known.

**Comment by maddyblue** [comment]
Ok. Handing responsibility to click the merge button to the cloud team. LGTM from my end.

File: src/balancerd/src/lib.rs
Comment (maddyblue) [reply src/balancerd/src/lib.rs:182]: Ah! Thanks, done.

**Review by alex-hunt-materialize — APPROVED** [review]
Code LGTM, but we'll need some branches in the pulumi code to handle the changed CLI args.

**Comment by alex-hunt-materialize** [comment]
Please note that we should never set the sigterm wait seconds higher than 600, as that is the pod's termination grace period, which cannot be set through launch darkly..

**Comment by jkosh44** [comment]
@alex-hunt-materialize I'm going to take over this PR on the adapter side from Maddy. My understanding is that it's good to go, but we're just waiting on the corresponding changes in cloud, is that accurate?

**Comment by benesch** [comment]
@jkosh44 looks like it's good to merge whenever. The [cloud PR](https://github.com/MaterializeInc/cloud/pull/9863) is prepared but there's no easy way to test it until this PR makes it into a release:

> I have only tested for the older version args, and don't have any easy way to test the new version until it is released. @maddyblue please check if these new args are reasonable.

So best thing to do is to merge this and then circle back next Thursday when this change makes it into the release so that https://github.com/MaterializeInc/cloud/pull/9863 can be tested against that release.

**Pull request merged**
